pkg/hubble/observer/namespace_manager.go

```diff
@@ -0,0 +1,95 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Hubble
+
+package observer
+
+import (
+	"context"
+	"time"
+
+	"golang.org/x/exp/slices"
+
+	observerpb "github.com/cilium/cilium/api/v1/observer"
+	"github.com/cilium/cilium/pkg/lock"
+)
+
+var _ NamespaceManager = &namespaceManager{}
+
+const (
+	checkNamespaceAgeFrequency = 5 * time.Minute
+	namespaceTTL               = time.Hour
+)
+
+type NamespaceManager interface {
+	GetNamespaces() []*observerpb.Namespace
+	AddNamespace(...*observerpb.Namespace)
+}
+
+type namespaceRecord struct {
+	namespace *observerpb.Namespace
+	added     time.Time
+}
+
+type namespaceManager struct {
+	mu         lock.RWMutex
+	namespaces map[string]namespaceRecord
+	nowFunc    func() time.Time
+}
+
+func NewNamespaceManager() *namespaceManager {
+	return &namespaceManager{
+		namespaces: make(map[string]namespaceRecord),
+		nowFunc:    time.Now,
+	}
+}
+
+func (m *namespaceManager) Run(ctx context.Context) {
+	ticker := time.NewTicker(checkNamespaceAgeFrequency)
+	defer ticker.Stop()
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-ticker.C:
+			// periodically remove any namespaces which haven't been seen in flows
+			// for the last hour
+			m.cleanupNamespaces()
+		}
+	}
+}
```

Comment:
@chancez Did you consider using `pkg/controller` in order to run this logic? The benefits are that it's easier to get visibility into when this periodic logic runs, whether there are errors, and so on. pkg/controller will automatically hook the logic into metrics as well as registering the status each time it runs with the `cilium status` reporter, so that users can understand whether the logic ran, when it ran, and whether it's stuck.

Reply:
No, but Hubble also is a bit of a snowflake in general, and hasn't had nearly as much focus as the rest of Cilium, including building better reusable primitives.